new database.yml format

arthurnn opened the pull request:
### Summary

New database.yml format. This is how it looks:
```ruby
default: &default
  adapter: sqlite3
  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
  timeout: 5000

development:
  primary:
    <<: *default
    database: db/development.sqlite3
  readonly:
    <<: *default
    database: db/readonly.sqlite3
```
We are moving from a two levels config to a three levels.

**This is backward compatible, apps using the default, old, 2 levels database.yml will still work.** 

### Advantages of this change
- Better structure on the database.yml file when dealing with multiples databases.
- One environment can have multiple configs.
- Having all the configs under an environment key, enables us to create/drop multiple databases on the create/drop tasks.
- Removes the necessity of adding the `Rails.env` when calling `establish_connection` for a non standard connection: i.e: `establish_connection "#{Rails.env}_readonly`

### Implementation details
The config file might have 3 levels, however, ActiveRecord now has a `local_configurations` option, which is scoped by the running `Rails.env`. That's why the code on `ConnectionSpecification::Resolver` didn't change much. Internally we still handle a 2 levels config Hash.
However, we are keeping the `configurations` getter/setter to be backward compatible and to be able to access it at the tasks level. (We might be able to deprecate this moving forward, that's another discussion tho)
The good thing about this, is that I was able to remove the `Rails.env` call from inside AR connection code. Now AR won't depend on the RAILS_ENV anymore.

### Caveats
If you were using a non-default `database.yml`, for instance:
```
development:
  adapter: sqlite3
development_readonly:
  adapter: sqlite3
```
There will be some necessary changes in the app, as AR will scope the config by the environment, so AR would only see:
```
development:
  adapter: sqlite3
```

The upgrade path would be to move the `development_readonly` config to inside `development` and reference it when calling `establish_connection`. (*see the third task on future work)

### Future work

- [ ] Change the guides.
- ~~Change the default database.yml template~~
- [ ] Connect to all databases under the env during Rails initializer, so apps won't need to call `establish_connection(:readonly)` for instance.
- [ ] Deprecate `create:all`/`drop:all` and make `create`/`drop` create all the dbs under the development/test envs?
- ~~Deprecate the old template/config with a warning message?~~

All the future work are not in this PR, because I would like to discuss one by one before implementing them, to make sure they make sense.


review @kaspth @dhh @rafaelfranca @tenderlove @jeremy 
@schneems (specially on commit https://github.com/rails/rails/commit/e51edd46a9cc1f4e0ab5ed6aac47c13a508940db, because it changes a lot from 6cc03675d30b58e28f585720dad14e947a57ff5b, However, I made sure the core functionality still works, which is make `DATABASE_URL` work without Rails.)

arthurnn committed:
Add local_configurations to AR

With .local_configurations we don't need to lookup the connection config
using an enviroment. This opens the door for us to have a 2 levels
config under an enviroment, to be able to make ActiveRecord work with
multiple connections.

arthurnn committed:
Refactoring on resolver test

arthurnn committed:
Move DATABASE_URL merge logic to config resolver

With the `DATABASE_URL` logic living in the connection config resolver, we don't
need to mess around with enviroment(such as RAILS_ENV, RACK_ENV) in it.
Because Active Record will now use the `local_configurations`, which is a
configuration hash for the running environment, the `DATABASE_URL` logic
gets easier to handle as it only needs to set the `url` config for the
`primary` database config, which is the default one.

arthurnn committed:
Remove RAILS_ENV from active record connection.

As we pass the local configuration to ActiveRecord, we don't need the
`Rails.env` anymore to find the right configuration.

arthurnn committed:
Move DATABASE_URL resolver test

arthurnn committed:
Move LegacyConfigTransformer to connection_spec file

also add tests

arthurnn committed:
Update broken namespace + tests

also make the default establish_connection call use :primary

arthurnn committed:
Remove RAILS_ENV call

arthurnn committed:
Update docs and remove un-used method.

arthurnn committed:
Adapt db rake tasks to use new yml config

arthurnn committed:
Use config and not env on connect code

arthurnn committed:
Fix broken tests

arthurnn committed:
Be more defensive on legacy config transformer

kirs commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,26 @@ def to_hash
 
+      # Converts a one level connection config, to the new format
+      # which is a two level, such as: {'primary' => {'adapter' => 'foo'}}
+      class LegacyConfigTransformer # :nodoc:
```
I don't think we should call it a legacy config. It's still a valid config that most of Rails apps are going to use, as long as we don't deprecate one level config.

matzke commented:
+1 for decoupling Rails.env call from inside AR connection code

-1 for 3 level configuration as a "default", because for simple projects setup is imho way too complex. if 3 levels work that's fine, but "default" should be the simple version

siegy22 commented:
Why do you use `ENV.fetch("RAILS_MAX_THREADS") { 5 }` instead of `ENV.fetch("RAILS_MAX_THREADS", 5)`? :grin:

dhh commented:
Like the new format, but concur with @matzke that most apps won't need multiple DBs starting out, so the default database.yml should still be a standard 2-level setup. Then we can explain the 3-level setup in the guide. So that informs that whatever we call the code isn't about "legacy", it's about 2-level depth or 3-level depth.

arthurnn commented:
👍  thanks @matzke and @dhh for the comment on the legacy naming.
I, indeed, thought about it multiple times. I agree that 98% of the rails apps won't probably need the 3 level connections, so let's not call it legacy, and make it live in our code base as a proper 'transformer' stage.

kaspth reviewed (commented):
Some initial rough thoughts and suggestions 😊

maclover7 commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -864,7 +864,7 @@ def connection_pool_list
 
       def establish_connection(config)
-        resolver = ConnectionSpecification::Resolver.new(Base.configurations)
+        resolver = ConnectionSpecification::Resolver.new(Base.local_configurations)
```
👍 agree with Kasper, the naming here is difficult to understand; it took me several read throughs to understand the difference between `local_configurations` and `configurations` 😬

arthurnn committed:
Make config work with 2 and 3 levels

Replace the LegacyTransformer, with a better configuration class. We
will still support 2 level configurations for apps that won't connect to
multiple databases, so we should not call legacy. Instead, we can
normalize the configuration before passing to the resolvers.

arthurnn commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -126,42 +156,35 @@ def initialize(configurations)
-        # Accepts a hash one layer deep that contains all connection information.
+        # Accepts:
+        # - Hash: one layer deep Hash that contains all connection information
+        # - Symbol: a connection specification name, that will be looked-up
```
👍  Agree @kaspth ,

So the model using a readonly config could look like this:

```ruby
class User < ApplicationRecord
  self.connection_configuration = :readonly
end
```

or:
```ruby
class User < ApplicationRecord
  def self.connection_configuration
    Application.readonly_mode? ? :readonly : :primary
  end
end
```

Pretty much we could replace the `connection_specification_name` to `connection_configuration` in the model.

I don't want to change those in this PR tho, we can try in a next one. Also there will be some deprecation cycle for the name change, as 5.0 uses `connection_specification_name`

arthurnn committed:
Update docs [skip ci]

arthurnn reviewed (commented):
I clean-up this a bit. There is no legacy reference anymore, as we will support 2 level configs moving forward.
@kaspth I liked your suggestion on having a Configuration class, that's what I did.
However, I didn't want to extend Hash, and instead, I did a Proxy. (http://words.steveklabnik.com/beware-subclassing-ruby-core-classes)
Also, I cannot normalize the config in the initializer, as `AR::Base.configurations` is a public API, and people could be using that configuration for something. so instead I just created a `normalize` method, that will return the config the way the resolver expects.

Thoughts?

arthurnn committed:
fix linter

arthurnn committed:
Fix db console

arthurnn committed:
Fix rubocop

arthurnn committed:
More rubocop fixes

arthurnn committed:
Make random top level config work on establish_connection

arthurnn commented:
@matthewd as discussed. I changed the PR to make this possible:
database.yml
```
production:
  database: ...
production_readonly:
  database: ...
```

So, old setups, with a 2 levels config, will still work, and will still be able to do a `establish_connection(:production_readonly)`

kirs commented:
Do we have any plan about how this works with migrations?

I see at least two scenarios:

1) You have `primary` and `readonly` connections and you want to use the first one to run the migrations.

2) You have different connections with different databases. In our case, `master` and `shard_X` databases have completely different tables. For every migration, we have a method that describes which database the migration is going to alter.

kaspth commented:
@kirs I think `bin/rails g migration some_name --db readonly` with `self.connection = :readonly` in the migration would suffice, but maybe we should shelve that discussion for later. Think this PR has big implications as is 😊

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,37 @@ def to_hash
         @config.merge(name: @name)
       end
 
+      class ConnectionConfigurations < SimpleDelegator #:nodoc:
```
Isn't inheriting from SimpleDelegator effectively the same thing as inheriting from Hash?

Maybe we should just restrict the allowed interface to `[]`?

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,37 @@ def to_hash
       end
 
+      class ConnectionConfigurations < SimpleDelegator #:nodoc:
+        attr_accessor :root_level
```
Since we're only using the instance variable there's no reason for the reader method. I'd do: `attr_writer`.

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,37 @@ def to_hash
+
+        def normalized(key = @root_level)
+          config = key ? self[key] : self
+          config = config ? config.dup : {}
```
Find these two lines hard to parse because of the two conditionals. The empty hash fallback seems too defensive at first sight. What case is it trying to cover? That there's a key in `self[key]` but with no config at all?

Could we simplify to this?

```ruby
config = fetch(key, self).dup
```

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,37 @@ def to_hash
+          end
+
+          if url = ENV["DATABASE_URL"]
+            config["primary"] ||= {}
```
We've just spent lines establishing `"primary"`, surely it should have been made an empty hash by now?

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -126,42 +159,35 @@ def initialize(configurations)
           if config
             resolve_connection config
-          elsif env = ActiveRecord::ConnectionHandling::RAILS_ENV.call
-            resolve_symbol_connection env.to_sym
```
Are we removing this because `resolve_connection` calls it?

arthurnn replied:
added back

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -126,42 +159,35 @@ def initialize(configurations)
-            config[key] = resolve(value) if value
-          end
-          config
-        end
```
This makes me wonder how much we need to worry about people calling all the available Hash methods on `ActiveRecord::Base.configurations[some_env]` and if we can just normalize the configurations up front.

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -223,19 +249,21 @@ def resolve_connection(spec)
           end
 
-          # Takes the environment such as +:production+ or +:development+.
+          # Takes the specification name such as +:primary+.
```
`configuration name` 😉

kaspth reviewed (commented):
Some more feedback ❤️

matthewd commented on activerecord/lib/active_record/connection_handling.rb:
```diff
@@ -44,10 +41,9 @@ module ConnectionHandling
     # The exceptions AdapterNotSpecified, AdapterNotFound and +ArgumentError+
     # may be returned on an error.
-    def establish_connection(config = nil)
+    def establish_connection(config = :primary)
```
I think `self.configuration` is way too vague to configure a model. In the bowels of AR, it's fine to infer that "configuration" means "Active Record configuration", and then assume that means "[database connection] configuration".

At the model layer, though, we should be more explicit: `User.configuration` is just unclear.

kaspth replied:
Agreed, it's too broad. Perhaps I should just learn to stop worrying and love `self.connection_specification_name =` 😄

sirupsen commented:
I like the new format as well, we have something like this supported by monkey-patches internally. I would share @kirs concerns about migrations. Does the schema cache do the right thing? That's an area where we'd be happy to get rid of our custom code.

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -126,42 +159,35 @@ def initialize(configurations)
-        #   config = { "production" => { "host" => "localhost", "database" => "foo", "adapter" => "sqlite3" } }
-        #   spec = Resolver.new(config).spec(:production)
+        #   config = { "host" => "localhost", "database" => "foo", "adapter" => "sqlite3" }
+        #   spec = Resolver.new({}).spec(config)
```
This API looks weird to me. I'd expect the resolver to always be initialized with the config and resolve a spec.

arthurnn replied:
fixed

rafaelfranca commented on activerecord/lib/active_record/railties/databases.rake:
```diff
@@ -324,22 +324,22 @@ db_namespace = namespace :db do
         should_reconnect = ActiveRecord::Base.connection_pool.active_connection?
         ActiveRecord::Schema.verbose = false
-        ActiveRecord::Tasks::DatabaseTasks.load_schema ActiveRecord::Base.configurations["test"], :ruby, ENV["SCHEMA"]
+        ActiveRecord::Tasks::DatabaseTasks.load_schema ActiveRecord::Tasks::DatabaseTasks.config_at("test"), :ruby, ENV["SCHEMA"]
```
Looks like it indeed broke.

arthurnn replied:
adding it back

rafaelfranca commented on activerecord/test/cases/tasks/database_tasks_test.rb:
```diff
@@ -152,23 +156,42 @@ def test_creates_configurations_with_blank_hosts
-        "development" => { "database" => "dev-db" },
-        "test"        => { "database" => "test-db" },
-        "production"  => { "database" => "prod-db" }
+        "development" => { "database" => "dev-db", "adapter" => "sqlite3" },
```
Why the adapter is now required?

kaspth replied:
I think it's because we only check for the `adapter` key at the moment. Finding a more accurate way to find an actual config could fix the need to change this.

matthewd replied:
It's already required in practice for an actual config you intend to use, and while this test was accidentally assuming it, we've never claimed to be an arbitrary hash storage bucket.

arthurnn committed:
Refactor connection configuration code

arthurnn committed:
Make establish_connection nil work with default config

arthurnn committed:
Fix comments

arthurnn committed:
revert some not needed changes

arthurnn committed:
Fix database_url merging code

kaspth commented:
Looks like the backwardscompatibility issues have been solved, so I'd say rebase, fix tests, run `rubocop -a`, squash and then you have my ok to merge 👍

This was a hard and long haul so I definitely appreciate your willingness to run my review gauntlet so many times 😄

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,45 @@ def to_hash
+          ret
+        end
+
+        def keys
```
When is this called? My cmd-E on `keys` can't find any instances?

kaspth commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,45 @@ def to_hash
+        def to_hash
+          @config
+        end
+      end
```
I still found a lot of this logic hard to read, so I took the liberty of trying to make it more concise:

```ruby
class ConnectionConfigurations #:nodoc:
  def initialize(config)
    @root_config = @config = config
    @database_url_config = resolve_database_url_config || {}
  end
 
  def scope=(scope)
    @root_config, @config = @config, (@config[scope] || {})
  end
 
  def [](key)
    @database_url_config.reverse_merge(@config[key] || @root_config[key] || {})
  end

  def keys
    @config.keys
  end

  def to_h
    @config
  end

  private
    def resolve_database_url_config
      if ENV["DATABASE_URL"]
        ConnectionUrlResolver.new(ENV["DATABASE_URL"]).to_hash
      end
    end
end
```

I'm curious why we need the `@root_config[key]` fallback in `[]`?

NOTE: `scope=` basically assumes that it's never assigned more than once, which it doesn't seem to be in library code, but it is in test code. Would it make more sense to pass it in at the constructor? Or have a `scope` method that returns a new instance of `ConnectionConfigurations`?

matthewd commented on activerecord/lib/active_record/connection_adapters/connection_specification.rb:
```diff
@@ -17,6 +17,45 @@ def to_hash
+          end
+          ret ||= @config[key]
+          if @database_url_config
+            ret = (ret || {}).merge(@database_url_config)
```
Isn't this applying the URL-supplied database to every configuration we return?

That's wrong, and I'm alarmed the tests aren't point that out. It's hard to see what's going on with them, because they're all renamed at the same time, though. 😟
